### PR TITLE
Auto ingest

### PR DESCRIPTION
**Pull request by dhkamra-microsoft**
Added an api to find the closest ingest server from a list of ingest servers. I have put it in ftl so any client that doesn't support curl can use it

### CODE REVIEW

File: libftl/ingest.c [hunk 1]
```diff
@@ -1,22 +1,150 @@
...
+    ingest->rtt = 1000;
+
+    if ((ping = _ping_server(ingest->ip, INGEST_PING_PORT)) >= 0) {
+        ingest->rtt = ping;
+    }
+
+    return 0;
+}
+
+ftl_status_t ftl_find_closest_ingest(const char* ingestIps[], const char* ingestNames[], int ingestsCount, char* bestIngestIpComputed)
```
Comment (QuinnDamerell-MS) [inline libftl/ingest.c]: I don't think this is used, so you might not need to pass it.

File: libftl/ingest.c [hunk 2]
```diff
@@ -1,22 +1,150 @@
...
+{
+    ftl_ingest_t* ingestElement;
+    int i;
+
+    if ((ingestElement = malloc(sizeof(ftl_ingest_t) * ingestsCount)) == NULL) {
+        return FTL_MALLOC_FAILURE;
+    }
+
+    for (i = 0; i < ingestsCount; i++)
+    {
```
Comment (QuinnDamerell-MS) [inline libftl/ingest.c]: nit: keep on same line as for

File: libftl/ingest.c [hunk 3]
```diff
@@ -1,22 +1,150 @@
...
+        strcpy_s(ingestElement[i].name, sizeof(ingestElement[i].name), ingestNames[i]);
+        strcpy_s(ingestElement[i].ip, sizeof(ingestElement[i].ip), ingestIps[i]);
+        ingestElement[i].rtt = 500;
```
Comment (QuinnDamerell-MS) [inline libftl/ingest.c]: might want to default this higher

File: libftl/ingest.c [hunk 4]
```diff
@@ -1,22 +1,150 @@
...
+    if ((ping = _ping_server(ingest->ip, INGEST_PING_PORT)) >= 0) {
+        ingest->rtt = ping;
+    }
+
+    return 0;
+}
+
+ftl_status_t ftl_find_closest_ingest(const char* ingestIps[], const char* ingestNames[], int ingestsCount, char* bestIngestIpComputed)
+{
+    ftl_ingest_t* ingestElement;
```
Comment (QuinnDamerell-MS) [inline libftl/ingest.c]: nit: elements

File: libftl/ingest.c [hunk 5]
```diff
@@ -1,22 +1,150 @@
...
+
+    for (i = 0; i < ingestsCount; i++)
+    {
+        strcpy_s(ingestElement[i].name, sizeof(ingestElement[i].name), ingestNames[i]);
+        strcpy_s(ingestElement[i].ip, sizeof(ingestElement[i].ip), ingestIps[i]);
+        ingestElement[i].rtt = 500;
+        ingestElement[i].next = NULL;
+    }
+
+    OS_THREAD_HANDLE *handle;
```
Comment (QuinnDamerell-MS) [inline libftl/ingest.c]: nit: handles

Comment (QuinnDamerell-MS) [reply libftl/ingest.c]: Actually, you could even add this to the thread_data below to make it a little cleaner.

File: libftl/ingest.c [hunk 6]
```diff
@@ -1,22 +1,150 @@
...
+    if ((data = (_tmp_ingest_thread_data_t *)malloc(sizeof(_tmp_ingest_thread_data_t) * ingestsCount)) == NULL) {
+        return FTL_MALLOC_FAILURE;
+    }
+
+    /*query all the ingests about cpu and rtt*/
+    for (i = 0; i < ingestsCount; i++) {
+        handle[i] = 0;
+        data[i].ingest = &ingestElement[i];
+        data[i].ftl = NULL; // This is not used by _ingest_get_rtt
+        os_create_thread(&handle[i], NULL, _ingest_get_rtt, &data[i]);
```
Comment (QuinnDamerell-MS) [inline libftl/ingest.c]: can this fail? if so should you abort?

File: libftl/ingest.c [hunk 7]
```diff
@@ -1,22 +1,150 @@
...
+    }
+
+    free(handle);
+    free(data);
+
+    if (best) {
+        strcpy_s(bestIngestIpComputed, sizeof(best->ip), best->ip);
+        return FTL_SUCCESS;
+    }
+
```
Comment (QuinnDamerell-MS) [inline libftl/ingest.c]: free(ingestElement)

Comment (QuinnDamerell-MS) [reply libftl/ingest.c]: well, do this before and after the return above

**Review by QuinnDamerell-MS — CHANGES_REQUESTED** [review]
Looks good overall, just a few little things. Thanks for the commit!

File: libftl/ingest.c (hunk 3 above)
Comment (dhkamra-microsoft) [reply libftl/ingest.c]: Using the same value as ftl uses inside _ingest_get_hosts.

File: libftl/ingest.c (hunk 6 above)
Comment (dhkamra-microsoft) [reply libftl/ingest.c]: Using the same logic as _ingest_get_hosts. I think its ok. Even if one of the threads fail, you can still recover. Although a thread creation failure probably means something much more malign.

**Review by dhkamra-microsoft — COMMENTED** [review]
Made changes as per quinn's comments. Added one small api after conferring with Stefan.

**Pull request merged**